src/libopensc/card-idprime.c

```diff
@@ -0,0 +1,698 @@
+	sc_path_t index_path;
+
+	/* First, we need to make sure the IDPrime AID is selected */
+	r = idprime_select_idprime(card);
+	if (r != SC_SUCCESS) {
+		LOG_FUNC_RETURN(card->ctx, r);
+	}
+
+	/* Returns FCI with expected length of data */
+	sc_format_path("0101", &index_path);
+	r = iso_ops->select_file(card, &index_path, &file);
+	if (r != SC_SUCCESS) {
+		sc_file_free(file);
+		LOG_FUNC_RETURN(card->ctx, r);
+	}
+	r = file->size;
+	sc_file_free(file);
+	return r;
+}
+
+static int idprime_process_index(sc_card_t *card, idprime_private_data_t *priv, int length)
+{
+	u8 *buf = NULL;
+	int r = SC_ERROR_OUT_OF_MEMORY;
+	int i, num_entries;
+	idprime_object_t new_object;
+
+	buf = malloc(length);
+	if (buf == NULL) {
+		goto done;
+	}
+
+	r = iso_ops->read_binary(card, 0, buf, length, 0);
+	if (r < 1) {
+		r = SC_ERROR_WRONG_LENGTH;
+		goto done;
+	}
+
+	/* First byte shows the number of entries, each of them 21 bytes long */
+	num_entries = buf[0];
+	if (r < num_entries*21 + 1) {
+		r = SC_ERROR_INVALID_DATA;
+		goto done;
+	}
+	new_object.fd = 0;
+	for (i = 0; i < num_entries; i++) {
+		u8 *start = &buf[i*21+1];
+
+		/* First two bytes specify the object DF */
+		new_object.df[0] = start[0];
+		new_object.df[1] = start[1];
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "df=%s",
+			sc_dump_hex(new_object.df, sizeof(new_object.df)));
+		/* I assume this is some identification of certificate */
+		if (start[4] == 0x6B && start[5] == 0x78 && start[6] == 0x63
+				&& start[7] == 0x30) {
+			new_object.fd++;
+			/* The key reference is one bigger than the value found here for some reason */
+			new_object.key_reference = start[8] + 1;
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Found certificate with fd=%d",
+				new_object.fd);
+			idprime_add_object_to_list(&priv->pki_list, &new_object);
+		}
+	}
+	r = SC_SUCCESS;
+done:
+	if (buf)
+		free(buf);
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+/*
+ * Look for a IDPrime card. If it exists, initialize our data structures
+ */
+static int idprime_find_and_initialize(sc_card_t *card, int initialize)
+{
+	int r;
+	idprime_private_data_t *priv = NULL;
+
+	/* already initialized? */
+	if (card->drv_data) {
+		return SC_SUCCESS;
+	}
+
+	r = idprime_select_index(card);
+	if (r > 0) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Index file found");
+		if (!initialize) {/* match card only */
+			return SC_SUCCESS;
+		}
+
+		priv = idprime_new_private_data();
+		if (!priv) {
+			return SC_ERROR_OUT_OF_MEMORY;
+		}
+		r = idprime_process_index(card, priv, r);
+		if (r == SC_SUCCESS) {
+			card->type = SC_CARD_TYPE_IDPRIME_GENERIC;
+			card->drv_data = priv;
+			return r;
+		}
+		idprime_free_private_data(priv);
+	}
+	return r;
+}
+
+static int idprime_init(sc_card_t *card)
+{
+	int r;
+	unsigned long flags;
+
+	/* TODO We should get this from the MF FCI:
+	 * 00 A4 04 00 00 (select MF returning FCI, tag 9F65 under tag A5 from GP
+	 */
+	card->max_send_size = (255 - 12);
+	card->max_recv_size = 255;
```

Comment:
not 256, i.e. Le=0 => Ne=256?

Reply:
Surprisingly this looks like working now even without this code so I will try to drop that for now.